WIP: Feature/dataframe aggregate (implements #1619)

chmp opened the pull request:
After the great experience with #1666, I started to implement `dd.DataFrame(...).groupby(...).agg(...)` on a recent, particularly long, train ride. The implementation is pretty much feature-complete. However, support for `var` and `std` is still missing. As of now, there are no changes to the existing dask code base for implementing this feature. Since this fact results in some code duplication and there are further design consideration to be discussed, I wanted to open a PR early to ask for feedback. I hope opening such a big issue as a PR without prior announcement is OK.

# Implementation notes

The implementation supports all documented argument values of pandas aggregate: 

- function (`.agg('mean')`
- list of functions (`.agg(['mean', 'sum'])`)
- dict of columns -> functions (`.agg({'b': 'mean', 'c': 'max'})`)
- nested dict of names -> dicts of functions (`.agg({'b': {'c': 'mean'}, 'c': {'a': 'max', 'a': 'min'}})`)

The aggregation functions `sum`, `min`, `max`, `count`, `size`, and `mean` are currently supported. When function objects are used, they are matched by name against known aggregates. Support for `std` and `var` is still missing.

I followed closely the implementation of the existing aggregate functions. However, I added an additional finalize-step to make it easy to support tree aggregations (since this topic seems to be discussed). This design results in three stages: chunk functions, aggregate functions, and finalizers. In `aggregate`, `aca` (chunk functions + aggregate functions) is followed by `map_partitions` (finalizers). 

Each stage applies multiple function objects to the (grouped) input dataframe and returns a fragment of the result dataframe. The full results is obtained by concatenating all fragments (`_groupby_apply_funcs` and `_agg_finalize`).

Most of the heavy lifting is performed by generating function objects that are applied to the dataframe and return fragments of the result dataframe (`_build_agg_args`). Since aggregate and finalize steps work on intermediate results of prior phases, `agg` forms an implicit expression graph. As of now, each aggregate generates its own independent graph by writing anonymous columns in intermediate dataframes (`next_id`). An alternative would be to use fixed names for intermediates and thereby reuse them between aggregates. For example, `.agg(['sum', 'mean', 'count'])` would only compute one sum and one count per input-column. 

Currently, all aggregations are implemented by calling individual aggregate functions on grouped dataframes (following existing code). Another option would be to call the pandas `agg` implementation to generate intermediates. Depending on the perfomance of the pandas implementation this change could improve overall performance.

Implements #1619.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+                                           next_id, simple_impl[func])
+
+        elif func == 'var':
+            raise NotImplementedError()
```
The functions in `dask/dataframe/groupby.py` like `_var_chunk` may be of value here.

chmp replied:
Thanks, for the pointer. I will probably reimplement it to be able to reuse intermediates across different reductions, if this is ok for you?

mrocklin replied:
As long as single-groupby-aggregation performance isn't seriously affected or made significanty more complex then, yes, modifying existing code sounds great.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+
+
+AggArgs = collections.namedtuple('AggArgs', ['chunk_funcs', 'aggregate_funcs',
+                                             'finalizers'])
```
Would a dictionary do ok here?  Most dask development has relatively few custom classes.  Namedtuples are pretty benign, I agree, but the majority of the codebase is pretty barebones.

chmp replied:
I'll replace the namedtuple with a dict.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+            partial(_apply_func_to_column, int_sum, int_sum, M.sum),
+            partial(_apply_func_to_column, int_count, int_count, M.sum)
+        ],
+        finalizers=[(result_column, partial(_finalize_mean, int_sum, int_count))],
```
There is a non-trivial cost to using `partial` functions in a distributed setting.  They are harder to serialize and harder to reason about when people need to debug the graphs themselves (which sometimes core developers do need to do).

This isn't a blocking request (fine if it's too hard) but if its possible to put these arguments directly into the graph rather than implicitly with a partial then that would be better.

chmp replied:
Thanks for the explanation. To be hones, I was wondering when reading the dask code base about this design choice, makes perfect sense. 

I'll put the arguments directly into the graph. My proposal would be to reverse the arguments to `_apply_func_to_column`, convert all partial arguments to keyword arguments, and to replace the partials with pairs of function and kwargs.

mrocklin commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -505,3 +505,35 @@ def test_groupby_multiprocessing():
+    assert eq(pdf.groupby('a').agg(spec), ddf.groupby('a').agg(spec))
+
+    spec = 'mean'
+    assert eq(pdf.groupby('a').agg(spec), ddf.groupby('a').agg(spec))
```
It might be nice to put all of the specs into a list and then loop over them.  Or perhaps use a `@pytest.parametrize` testing fixture.

chmp replied:
I will go with `pytest.parametrize` (and fix the failing tests due to the recent changes on master).

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -122,6 +126,185 @@ def _nunique_series_aggregate(df):
+        'max': (M.max, M.max),
+        'count': (M.count, M.sum),
+        'size': (M.size, M.sum),
+    }
```
You might want to see the very recent work of tree reductions here: https://github.com/dask/dask/pull/1663

This ends up having a significant performance impact for some groupbys.  It would be nice (eventually) to have the same logic in `aggregate`.  This would require adding another intermediate combine function to each of these tuples.

mrocklin replied:
OK

chmp replied:
For all aggregation supported `combine == aggregate`. While, I would be happy to add the corresponding tuple elements, dictionary entries in the `_build_ ...` functions, I feel it would needlessly clutter the code. However, I think adding the corresponding keyword arguments for `apply_concat_apply` would make the assumption of `combine == aggregate` explicit.

chmp commented:
I rebased to master and addressed your comments (namedtuples, partial, pytest parametrize). Further, I changed the semantics of the the funcs argument in `_groupby_apply_funcs`. Now, each function returns exactly one column. This changed allowed me to skip recomputing intermediate sums. For certain `aggregate` calls this should reduce the communication / computation overhead by a factor of two or more. A specific example would be `df.groupby(...).agg(['mean', 'sum', 'count'])`. Finally, I also implemented `var` and `std`.

Next, I will takle tree reductions.

chmp commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +136,243 @@ def _nunique_series_aggregate(df):
+    # generated it is reused in subsequent id requests.
+    ids = it.count()
+    # TODO: add func, input_column to intermediate name for easier debugging
+    id_map = collections.defaultdict(lambda: 'intermediate-{}'.format(next(ids)))
```
Currently, the skipping of recomputation is implemented by a somewhat arcane combination of `itertools.count` and `defaultdict`. What do you think about implementing it as a separate class? This would also make it simpler to generate easy to interpret intermediate names.

mrocklin replied:
Are these names used as keys in the task graph or just as column names?  Is there a convenient deterministic way to create these names?

Generally I avoid classes, probably somewhat irrationally.

chmp replied:
Rightly so: now that I think about it, this functionality can also be easily implemented with a closure. 

The keys are only used as column names in intermediate data frames. 

With regard to generating these names deterministically. I am somewhat certain that prefixing the repr of the column key by the function should be a unique key, as long as the separator is not used inside the function name. However, I thought better to make sure by adding a unique component to the name.

A possible implementation could be:

```python
def _make_agg_id_factory():
    integers = it.count()
    ids = {}
    
    def factory(func, column):
        try:
            return ids[func, column]

        except KeyError:
            new_id = '{}-{!s}-{!r}'.format(next(integers), func, column)
            return ids.setdefault((func, column), new_id)

    return factory
```

chmp commented:
Regarding the failed travis check: I cannot' see any place where I would have changed code that is relevant to that test. Are there any issues known?

chmp commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -553,3 +557,71 @@ def test_groupby_multiprocessing():
+
+    assert_eq(pdf.groupby(['a', 'd']).agg(spec),
+              ddf.groupby(['a', 'd']).agg(spec))
+
```
After adapting the tests. I found a couple of problems:

1. using series as a groupby argument fails miserably
2. pandas is using additional magic where it tries to cast the result back to its original type. The relevant code can be found in [_GroupBy._try_cast](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby.py#L760). (see also #1685). While it is definitely possible to ensure the final data frame has the correct types after `.compute()`. This would also mean, that there is no way of knowing the correct meta object before performing the actual aggregation
3. `.groupby(...).agg('size')` returns a series for inconsistence sake

chmp replied:
My recent commits address all issues 1 and 3. For Issue 2 however, i.e., the recasting of the result, I don't see any way to fix it.

mrocklin commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -553,3 +555,57 @@ def test_groupby_multiprocessing():
+    assert len(no_mean_aggs) == len(with_mean_aggs)
+
+    assert len(no_mean_finalizers) == len(no_mean_spec)
+    assert len(with_mean_finalizers) == len(with_mean_spec)
```
> Check 3 should always be satisfied. Irrespective of the chosen aggregations only a single dataframe will be generated per partition. Therefore, the number of elements inside the dask graphs should always be same.

Very cool.  Besides being more convenient I anticipate that this will have some very nice performance benefits :)

chmp commented on dask/dataframe/groupby.py:
```diff
@@ -411,6 +702,63 @@ def __getattr__(self, key):
+        else:
+            levels = 0
+
+        # TODO: add normed spec as an additional part to the token?
```
Are there any specifics I need to watch out when generating the token or does apply-concat-apply ensures that the resulting dask keys are unique?

chmp commented on dask/dataframe/groupby.py:
```diff
@@ -150,18 +452,7 @@ def __init__(self, df, index=None, slice=None, **kwargs):
-                index = list(index.columns)
-
-        self.index = index
+        self.index = _normalize_index(df, index)
```
Here, I added support for lists of series as a grouper, which was previously not supported. This logic is required to handle `df.groupby([df['a'], df['b']])` consistent with pandas. The recursion in `_normalize_index` is not truly necessary (since lists of lists should not be valid groupers), but it was easiest way to implement it.

chmp commented on dask/dataframe/groupby.py:
```diff
@@ -411,6 +702,63 @@ def __getattr__(self, key):
+        meta = _aggregate_meta(self.obj._meta, meta_groupby, 0, chunk_funcs,
+                               aggregate_funcs, finalizers)
+
+        if not isinstance(self.index, list):
```
Here, I'm adding multiple series as varargs to the cunk functions. This way no additional logic for aligning the different series is required. The chunk functions have been adapted accordingly.

chmp commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -553,3 +558,143 @@ def test_groupby_multiprocessing():
+        agg_dask1 = get_agg_dask(result1)
+        agg_dask2 = get_agg_dask(result2)
+
+        core_agg_dask1 = {k: v for (k, v) in agg_dask1.dask.items()
```
`map_partitions` used in the finalize step passes the meta dataframe. Therefore, `==` can no longer be used for comparison.

mrocklin commented:
In travis.ci it looks like there is an encoding issue with one of the files and some flake8 issues generally.  Nothing major though.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -411,6 +702,63 @@ def __getattr__(self, key):
+
+        elif isinstance(self.index, list):
+            group_columns = {i for i in self.index
+                             if isinstance(i, tuple) or np.isscalar(i)}
```
Does the input order not matter here?

chmp replied:
The `group_columns` columns are only used to create the list of `non_group_columns`. Since only `contains` is checked the order of `group_columns` is relevant.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -411,6 +702,63 @@ def __getattr__(self, key):
+
+        # TODO: add normed spec as an additional part to the token?
+        token = 'aggregate'
+        finalize_token = '{}-finalize'.format(token)
```
Also, it might be nicer to just inline 'aggregate-finalize' directly in the `map_partitions` call.  

Removing small logic like this should make it easier for other developers to maintain this function in the future.

chmp replied:
I assumed additional logic was required. Since this is not the case, I removed both tokens and inlined them directly into the `aca` and `map_partitions` calls.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +136,304 @@ def _nunique_series_aggregate(df):
+
+def _normalize_spec(spec, non_group_columns):
+    """
+    Return a list of ``(result_column, func, input_column)`` tuples.
```
Can I ask for a quick `Examples` section here with a couple of worked examples to help future developers quickly understand how this function operates.  

```
Examples
--------

>>> _normalize_spec(simple, inputs)
... simple outputs ...
```

chmp replied:
I added further explanation to the docstring and different examples that should cover the functionality.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +136,304 @@ def _nunique_series_aggregate(df):
+#
+###############################################################
+
+def _make_agg_id_factory():
```
Should this function be tested independently?

chmp replied:
I added an explicit test.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +136,304 @@ def _nunique_series_aggregate(df):
+
+def _groupby_apply_funcs(df, *index, **kwargs):
+    """
+    TODO: document args
```
Lingering TODO here

chmp replied:
Add full description of this function with all its arguments.

chmp commented on dask/dataframe/groupby.py:
```diff
@@ -411,6 +702,63 @@ def __getattr__(self, key):
+
+        meta_groupby = pd.Series([], dtype=bool, index=self.obj._meta.index)
+        meta = _aggregate_meta(self.obj._meta, meta_groupby, 0, chunk_funcs,
+                               aggregate_funcs, finalizers)
```
I inlined it.

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -174,15 +174,51 @@ def factory(func, column):
+    --------
+
+        >>> _normalize_spec('mean', ['a', 'b', 'c'])
+        [('a', 'mean', 'a'), ('b', 'mean', 'b'), ('c', 'mean', 'c')]
```
These should be unindented to line up with the normal text on the left.  We should also drop the newline after the `----`.

http://dask.readthedocs.io/en/latest/develop.html#docstrings

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -174,15 +174,51 @@ def factory(func, column):
+        ...                  'b': {'e': 'count', 'f': 'var'}},
+        ...                 ['a', 'b', 'c'])
+        [(('a', 'mean'), 'mean', 'a'), (('a', 'size'), 'size', 'a'),
+         (('b', 'e'), 'count', 'b'), (('b', 'f'), 'var', 'b')]
```
These are super-informative by the way.  Thanks!

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -360,7 +396,31 @@ def _build_agg_args_mean(result_column, func, input_column, id_factory):
+    funcs: list of result-colum, function, keywordargument triples
+        The list of functions that are applied on the grouped data frame.
+        Has to be passed as a keyword argument.
+
```
Should drop endlines here

mrocklin commented:
This is looking pretty good to me.  It would be nice to have someone more familiar with dask.dataframe take a look at it before merging.  @jcrist if you have time tomorrow this could use your review.

jreback commented:
first (test) usage, seemed to work nicely.

intentional to support *only* DataFrameGroupby ATM? (fine by me), maybe put in a ``NotImplementedError`` on ``SeriesGroupby`` usage though)

jreback commented:
```
In [13]: ddf = dd.from_pandas(pd.DataFrame({'A' : [1,2,3], 'B' : [1, 2, 3]}), npartitions=3)

In [14]: ddf
Out[14]: dd.DataFrame<from_pa..., npartitions=2, divisions=(0, 1, 2)>

In [15]: ddf.compute()
Out[15]:
   A  B
0  1  1
1  2  2
2  3  3

In [16]: ddf.groupby('A').sum()
Out[16]: dd.DataFrame<datafra..., npartitions=1>

In [17]: ddf.groupby('A').sum().compute()
Out[17]:
   B
A
1  1
2  2
3  3

In [18]: ddf2 = ddf.groupby('A').sum()

In [19]: ddf2.columns = ['B2']

In [20]: ddf2
Out[20]: dd.DataFrame<rename-..., npartitions=1>

In [21]: ddf2.compute()
Out[21]:
   B2
A
1   1
2   2
3   3

In [22]: ddf2 = ddf.groupby('A').agg(['sum', 'mean'])

In [23]: ddf2
Out[23]: dd.DataFrame<aggrega..., npartitions=1>

In [24]: ddf2.compute()
Out[24]:
    B
  sum mean
A
1   1  1.0
2   2  2.0
3   3  3.0

In [25]: ddf2.columns = ['foo', 'bar']

In [26]: ddf2.compute()
Out[26]:
    B
  sum mean
A
1   1  1.0
2   2  2.0
3   3  3.0

In [27]: ddf = dd.from_pandas(pd.DataFrame({'A' : [1,2,3], 'B' : [1, 2, 3]}), npartitions=3)

In [28]: ddf
Out[28]: dd.DataFrame<from_pa..., npartitions=2, divisions=(0, 1, 2)>

In [29]: ddf.compute()
Out[29]:
   A  B
0  1  1
1  2  2
2  3  3

In [30]: ddf2 = ddf.groupby('A').agg(['sum', 'mean'])

In [31]: ddf2.compute()
Out[31]:
    B
  sum mean
A
1   1  1.0
2   2  2.0
3   3  3.0

In [32]: ddf2.columns = ['foo', 'bar']

In [33]: ddf2.compute()
Out[33]:
    B
  sum mean
A
1   1  1.0
2   2  2.0
3   3  3.0

In [35]: import dask

In [36]: dask.__version__
Out[36]: '0.11.1+54.g7f411f8'
```

chmp commented:
Regarding the missing aggregate for `Series`: this is a mere oversight, since it did came up in my use case. It should be easy to add, I'll have a look at it.

jcrist commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +136,363 @@ def _nunique_series_aggregate(df):
+    integers = it.count()
+    ids = {}
+
+    def factory(func, column):
```
Why not just use `dask.base.tokenize`? This provides deterministic keys, without the need for cache.

```
tokenize(func, column)
```

chmp commented:
The rename is not related to my code. dask does not seem to handle multilevel columns too well (for example also df['A', 'mean] does not work). A failing test for the rename functionality could be:

```python
def test_renames_multilevelcolumns():
    pdf = pd.DataFrame({('A', '0') : [1,2,2], ('B', 1) : [1, 2, 3]})
    ddf = dd.from_pandas(pdf, npartitions=3)

    pdf.columns = ['foo', 'bar']
    ddf.columns = ['foo', 'bar']

    assert_eq(pdf, ddf)

# E       AssertionError: Index are different
# E       
# E       Index classes are not equivalent
# E       [left]:  Index(['foo', 'bar'], dtype='object')
# E       [right]: MultiIndex(levels=[['A', 'B'], [1, '0']],
# E                  labels=[[0, 1], [1, 0]])
```

The culprit is the `dask.dataframe.core._rename ` implementation:

```python
def test_renames_pandas():
    pdf = pd.DataFrame({('A', '0') : [1,2,2], ('B', '1') : [1, 2, 3]})

    expected = pdf.copy()
    expected.columns = ['foo', 'bar']

    actual = pdf.rename(columns=dict(zip(pdf.columns, ['foo', 'bar'])))

    assert_eq(expected, actual)

# result:
# E       AssertionError: DataFrame.columns are different
# E       
# E       DataFrame.columns classes are not equivalent
# E       [left]:  Index(['foo', 'bar'], dtype='object')
# E       [right]: MultiIndex(levels=[['A', 'B'], ['0', '1']],
# E                  labels=[[0, 1], [0, 1]])
```

jreback commented:
@chmp in pandas you do this as ``df.loc[('A', 'mean')]``

chmp commented:
I posted a comment on the `__getitem__` issue.

mrocklin commented:
I think that we can safely ignore the multi-named columns issue for this PR.  It looks to me like the only outstanding issue it to use tokenize rather than the factory naming solution.

Any other comments?

chmp commented:
My most recent change addresses the two open comments:

- The factory for intermediate ids is gone. The unique part of the id comes now from `tokenize`.
- Series aggregate is implemented. 

However, currently series-aggregate does not work with multiple groupers, i.e., `s.groupby([s1, s2]).aggregate(...)`. The reason is a limitation of the groupby implementation for series in dask master: when I comment out the check in `SeriesGroupby.__init__`, the short-circuit branches fail, whereas the full aggregate code works as intended. I could have a look as this issue as part of the PR, however I would prefer to open a separate issue to address this topic. 

For reference when I remove all checks in `SeriesGroupy.__init__`, the test results are

```
dask/dataframe/tests/test_groupby.py::test_series_aggregate__examples[grouper0-None-spec0] XPASS
dask/dataframe/tests/test_groupby.py::test_series_aggregate__examples[grouper0-None-spec1] XPASS
dask/dataframe/tests/test_groupby.py::test_series_aggregate__examples[grouper0-None-spec2] XPASS
dask/dataframe/tests/test_groupby.py::test_series_aggregate__examples[grouper0-None-sum] xfail
dask/dataframe/tests/test_groupby.py::test_series_aggregate__examples[grouper0-None-size] xfail
```

mrocklin commented on dask/dataframe/groupby.py:
```diff
@@ -132,6 +137,340 @@ def _nunique_series_aggregate(df):
+#
+###############################################################
+def _make_agg_id(func, column):
+    return '{!s}-{!s}-{}'.format(func, column, tokenize(func, column))
```
Are func and column always strings?  If so then perhaps just `func + '-' + column`?

chmp replied:
Not necessarily. In principle columns can be any valid python type. In particular integers, booleans, and tuples come to my mind.

mrocklin commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -554,3 +556,169 @@ def test_groupby_multiprocessing():
+    ddf = dd.from_pandas(pdf, npartitions=10)
+
+    assert_eq(pdf.groupby(grouper(pdf)).agg(spec),
+              ddf.groupby(grouper(ddf)).agg(spec, split_every=split_every))
```
Nice test!

chmp replied:
Thanks. After your little nudge towards `pytest.mark.parametrize`, I became I fan of it :)

mrocklin commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -554,3 +556,169 @@ def test_groupby_multiprocessing():
+    ['sum', 'mean', 'min', 'max', 'count', 'size', 'std', 'var'],
+    'sum', 'size',
+])
+@pytest.mark.parametrize('split_every', [None])#[False, None])
```
Why not test `split_every = False` here?

chmp replied:
An oversight: I commented it out to increase the test speed and forgot to comment it in. I'll run the travis tests on a branch in my repo and push the changes after everything is green.

mrocklin commented:
Feel free to use travis.ci here.  There isn't a big need to isolate your
testing if you don't find it convenient.

On Tue, Oct 25, 2016 at 12:32 PM, Christopher Prohm <
notifications@github.com> wrote:

> *@chmp* commented on this pull request.
> ------------------------------
>
> In dask/dataframe/tests/test_groupby.py
> <https://github.com/dask/dask/pull/1678>:
>
> > +                        'c': [0, 1, 2, 3, 4, 5, 6, 7, 8] * 10,
> +                        'd': [3, 2, 1, 3, 2, 1, 2, 6, 4] * 10},
> +                       columns=['c', 'b', 'a', 'd'])
> +    ddf = dd.from_pandas(pdf, npartitions=10)
> +
> +    assert_eq(pdf.groupby(grouper(pdf)).agg(spec),
> +              ddf.groupby(grouper(ddf)).agg(spec, split_every=split_every))
> +
> +
> +@pytest.mark.parametrize('spec', [
> +    {'b': 'sum', 'c': 'min', 'd': 'max'},
> +    ['sum'],
> +    ['sum', 'mean', 'min', 'max', 'count', 'size', 'std', 'var'],
> +    'sum', 'size',
> +])
> +@pytest.mark.parametrize('split_every', [None])#[False, None])
>
> An oversight: I commented it out to increase the test speed and forgot to
> comment it in. I'll run the travis tests on a branch in my repo and push
> the changes after everything is green.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/1678>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASszG-I3iFmr2neQlS3EXGrHBcpd5htks5q3i8ygaJpZM4KaWX0>
> .
>

chmp commented:
After I enabled travis for my fork, I only have to push into a branch that is not connected to this PR. Then travis is running the tests isolated from the dask repo. Since, dask is already integrated with travis, this whole procedure is very convenient :)

mrocklin commented:
Any further comments on this PR?

mrocklin commented:
Alrighty.  Merging.  Thanks for all of the effort on this one @chmp !  Nice work.
